src/request.rs

```diff
@@ -1,220 +1,390 @@
 use crate::errors::{Error, MeilisearchError};
+use async_trait::async_trait;
 use log::{error, trace, warn};
 use serde::{de::DeserializeOwned, Serialize};
 use serde_json::{from_str, to_string};
 
 #[derive(Debug)]
-pub(crate) enum Method<Q, B> {
+pub enum Method<Q, B> {
     Get { query: Q },
     Post { query: Q, body: B },
     Patch { query: Q, body: B },
     Put { query: Q, body: B },
     Delete { query: Q },
 }
 
-#[cfg(not(target_arch = "wasm32"))]
-pub fn add_query_parameters<Query: Serialize>(url: &str, query: &Query) -> Result<String, Error> {
-    let query = yaup::to_string(query)?;
+#[async_trait(?Send)]
+pub trait HttpClient: Clone + Serialize + Send + Sync {
+    #[cfg(not(target_arch = "wasm32"))]
+    async fn request<Query, Body, Output>(
+        self,
+        url: &str,
+        apikey: Option<&str>,
+        method: Method<Query, Body>,
+        expected_status_code: u16,
+    ) -> Result<Output, Error>
+    where
+        Query: Serialize + Send + Sync,
+        Body: Serialize + Send + Sync,
+        Output: DeserializeOwned + 'static + Send;
+
+    #[cfg(target_arch = "wasm32")]
+    async fn request<Query, Body, Output>(
+        self,
+        url: &str,
+        apikey: Option<&str>,
+        method: Method<Query, Body>,
+        expected_status_code: u16,
+    ) -> Result<Output, Error>
+    where
+        Query: Serialize + Send + Sync,
+        Body: Serialize + Send + Sync,
+        Output: DeserializeOwned + 'static + Send;
+
+    #[cfg(not(target_arch = "wasm32"))]
+    async fn stream_request<
+        'a,
+        Query: Serialize + Send + Sync,
+        Body: futures_io::AsyncRead + Send + Sync + 'static,
+        Output: DeserializeOwned + 'static,
+    >(
+        self,
+        url: &str,
+        apikey: Option<&str>,
+        method: Method<Query, Body>,
+        content_type: &str,
+        expected_status_code: u16,
+    ) -> Result<Output, Error>;
+}
 
-    if query.is_empty() {
-        Ok(url.to_string())
-    } else {
-        Ok(format!("{url}?{query}"))
+#[derive(Debug, Clone, Serialize)]
+pub struct IsahcClient;
+
+impl IsahcClient {
+    pub fn new() -> Self {
+        IsahcClient
     }
 }
 
-#[cfg(not(target_arch = "wasm32"))]
-pub(crate) async fn request<
-    Query: Serialize,
-    Body: Serialize,
-    Output: DeserializeOwned + 'static,
->(
-    url: &str,
-    apikey: Option<&str>,
-    method: Method<Query, Body>,
-    expected_status_code: u16,
-) -> Result<Output, Error> {
-    use isahc::http::header;
-    use isahc::http::method::Method as HttpMethod;
-    use isahc::*;
-
-    let builder = Request::builder().header(header::USER_AGENT, qualified_version());
-    let builder = match apikey {
-        Some(apikey) => builder.header(header::AUTHORIZATION, format!("Bearer {apikey}")),
-        None => builder,
-    };
+impl Default for IsahcClient {
+    fn default() -> Self {
+        Self::new()
+    }
+}
```

Comment:
```suggestion
```

Since we can derive `Default` we don’t need this anymore

Reply:
fixed
https://github.com/meilisearch/meilisearch-rust/pull/459/commits/e7e2c355b1e410a02cb81d06cb4d9fa0f6ec9b9c